### DISCUSSION TITLE
Collection Literals and Constructor Parameters

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#7622] [Language Ideas]
Hi C# team:
I use HashSet and dictionaries s a lot in my code and would love to be able to use them with collection literals.  They work great when I don't need an ```IComparer<T>``` but once I do, I can't use them.

I want to throw out an idea on how IComparers could be used.

Allow calls with a syntax like one of the following:
Option 1: ```Test(["1", "2", "3"] with { comparer = StringComparer.InvariantCulture, });```
Option 2: ```Test(using { comparer = StringComparer.InvariantCulture} ["1", "2", "3"]);```
Option 3: ```Test(with { comparer = StringComparer.InvariantCulture} ["1", "2", "3"]);```

(Option 3 is my preference)

In the code above, ```with/using``` means "When selecting the constructor for the specified collection literal, the constructor that is selected by the compiler must have a parameter named ```comparer``` which is compatible with the value provided.  In the event that multiple constructors meet all of the specified criteria, the compiler is free to pick any of the available constructors."  The syntax would essentially be like partially specifying a PowerShell ```argset```.  Also, while I might only be specifying the ```comparer``` argument, any number of constructor parameters could be specified.

**Comment by HaloFour** [comment]
Doesn't seem to be much of a difference using existing syntax with target-typed new:

```cs
Test(new(StringComparer. InvariantCulture) { "a", "b", "c"} );
```

**Reply by TonyValenti** [reply]
You are correct but my understanding is that the collection initializers are not gaining all the performance enhancements made for collection literals.

**Reply by KennethHoff** [reply]
In addition to what Tony's saying; some collections - like immutable arrays - literally don't work with initilizers.

**Reply by qrli** [reply]
For those collections with constructor taking an initialization collection, such as HashSet<T>, the following also works:
```cs
Test(new(["a", "b", "c"], StringComparer.InvariantCulture) );
```

**Comment by ufcpp** [comment]
If we were talking only about collection expressions, I would be against this idea, but if it could be used for string interpolation culture propagation, etc., it might be a little more appealing.

**Reply by jnm2** [reply]
I don't share the full position here, but I do find the extension to interpolated strings interesting. Only thing is that the interpolated string case can be a simple API without language support.

**Comment by jnm2** [comment]
`with { capacity = 42 }` could help in those situations where the collection will be modified after construction and there's no way for the compiler to plan for that.

**Comment by qrli** [comment]
Why not just enable collection literal in the position of initializer?
```cs
Test(new(StringComparer.InvariantCulture) [ "a", "b", "c" ] );
```

**Reply by FaustVX** [reply]
Yes please. This will allow spread elements collection, which is not possible with collection initializer.

**Reply by HaloFour** [reply]
The following syntax is already legal, as you can immediately index off of a construction expression:

```cs
var x = new Foo("bar") [ 1, 2, 3 ];
```

You can argue that such syntax does not work with target-typed new, but I would not expect omitting the type name to drastically change the behavior.

**Reply by qrli** [reply]
My key point is that: rather than introducing new syntax at the constructor parameter side (where it can be much more complicated), it is better to do something at the position of initializer. The syntax is up to discussion. E.g. a silly idea:
```cs
Test(new(StringComparer.InvariantCulture) using [ "a", "b", "c" ] );
```

**Reply by TahirAhmadov** [reply]
@qrli what if the target type is `IEnumerable<T>`, then `new(...)` doesn't make sense, so C# has to do a look-ahead, which is more complicated than it needs to be.

**Reply by qrli** [reply]
> what if the target type is IEnumerable<T>, then new(...) doesn't make sense,

In that case, there is no need for extra parameters to pass.

**Reply by TahirAhmadov** [reply]
True, there is no need, but the language still needs to handle this from interpreting the tokens perspective, which adds complexity.

**Comment by iam3yal** [comment]
Another alternative that might work is pass the arguments to the constructor as part of the collection expression itself so something like this:
```
[(StringComparer.InvariantCulture), "1", "2", "3"]
```
I'm not sure whether it's confusing so maybe something like this:
```
[(StringComparer.InvariantCulture), ["a", "b", "c"]]
```

**Reply by KennethHoff** [reply]
I really like that. It's somewhat going against the "One size fits all" premise of collection expressions, but the problem this solves necessitates a deviation from that premise.

**Reply by HaloFour** [reply]
Those are both legal collections literals.  Reinterpreting the first element, or any of the elements, as members of an arbitrary constructor would pose some massive overload issues.  For example, most of those types also have a constructor that accepts an `int` to denote capacity, does that mean if you tried to create such a collection of ints using a literal that what you believe to be the first element would instead always be interpreted as the argument to that constructor?

I think a much more intentional syntax is necessary here, for the sake of clarity and disambiguation.

**Reply by iam3yal** [reply]
@HaloFour 

> does that mean if you tried to create such a collection of ints using a literal that what you believe to be the first element would instead always be interpreted as the argument to that constructor?

No, only if it's wrapped inside parentheses. I forgot to add parentheses in the 2nd example, sorry.

**Reply by HaloFour** [reply]
> No, only if it's wrapped inside parentheses.

Wrapping an expression in parenthesis is already legal, and doesn't change it's meaning.

```cs
int[] x = [(1), ((2)), ((((3))))];
```

**Reply by En3Tho** [reply]
What about syntax like `[values...](args...)`
```csharp
HashSet<int> hs = [ 1, 2, 3 ] (MyComparer.Instance)
```
It will require extending support for `Create(ReadOnlySpan values..., args...)` collection builder method

Named arguments can be supported too like
```csharp
List<int> values = [ 1, 2, 3 ] (capacity = 42)
```

**Reply by iam3yal** [reply]
Another option that I think reads well is something like this:
```
["a", "b", "c"] with (StringComparer.InvariantCulture)
```

**Reply by KennethHoff** [reply]
if/when https://github.com/dotnet/csharplang/issues/162 releases, could this be problematic?

It uses `(...)` instead of `{...}`, but it could still be loosely conflicting - if not only for the eyes of the developers.

Let's say you have a `MyCollection` that has a `with pattern` that allows you to set the `MaximumElements` _and_ a constructor with a StringComparer 

```c#
MyCollection<int> = [0, 1, 2] with (StringComparer.InvariantCulture) with { MaximumElements = 10 };
```

It's kind of a mess.

Thinking some more; Still wouldn't even need https://github.com/dotnet/csharplang/issues/162 as `MyCollection` could simply be a `record` or a `struct` (or a `record struct`) - which already supports the `with` operator.

**Reply by iam3yal** [reply]
@KennethHoff I doubt it. Why do you think so? 

> Thinking some more; Still wouldn't even need https://github.com/dotnet/csharplang/issues/162 as MyCollection could simply be a record or a struct (or a record struct) - which already supports the with operator.

Yes but then it uses bracers as opposed to parentheses.

**Reply by KennethHoff** [reply]
> @KennethHoff I doubt it. Why do you think so? 

I updated my comment for clarity

**Reply by CyrusNajmabadi** [reply]
It feels to me that the easy way to do this will just be extension methods on collection literals, which we want to do in c#13.  

So you'd just do `[x, y, ..z].ToSet(comparer)`.  Simple, clean, readable. Performance would normally be good, with stack allocing the temp values to pass in, etc.

This feels like a "good enough" solution to this corner case.

**Reply by HaloFour** [reply]
> It feels to me that the easy way to do this will just be extension methods on collection literals

Has the added benefit that the extension methods could be intentionally written to support collection literals, whereas some constructors probably don't make a lot of sense.

**Reply by iam3yal** [reply]
What is the syntax? something like this `ToSet(this [] source, ...)`?

**Reply by HaloFour** [reply]
I think the type would be `Span<T>` or `ReadOnlySpan<T>`?

**Reply by iam3yal** [reply]
ahh right! makes sense.
